ntfs_forensics/src/ntfsfilenameattributecontents.cpp

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2018 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ntfsfilenameattributecontents.h"
+
+namespace trailofbits {
+namespace {
+uint32_t unixtimestamp(uint64_t ntdate) {
+#define NSEC_BTWN_1601_1970 (uint64_t)(116444736000000000ULL)
+
+  ntdate -= (uint64_t)NSEC_BTWN_1601_1970;
+  ntdate /= (uint64_t)10000000;
+
+  return (uint32_t)ntdate;
+}
+}
+
+bool NTFSFileNameAttributeContents::valid() const {

```

Comment:
I'm not sure yet, maybe @GarretReece can shed some light on it

Reply:
There aren't many ways to check whether a recovered attribute block is valid, so the approach here is to check whether the timestamps all look reasonably sane. If all the timestamps fall between 1970-01-01-00:00:00.00 and 2025-01-01-00:00:00.0000 then we declare the file name attributes block to be valid.

Reply:
What's the plan for after the year 2025 though?

Reply:
Charge $4k for a patch every year.

Or we could bump the valid range, that's an option too.

Reply:
What if instead of that, we leave off the upper bound, and let the user decide if they're looking at valid timestamps? 

Ideally we might like to simply check if timestamps were in the future, but there's no good way to do that (without checking an authoritative time source, making sure there were no timezone or daylight savings time issues, etc.).

Reply:
I think we should report the raw value, and have the log aggregator decide how to validate it (mostly for the reason that Mike has outlined)

Reply:
this check is used when recovering entries from the slack space in directory indices--the approach is to walk to the slack space byte by byte, checking if at that offset into the slack whether the data in that space looks like a filename attribute entry. It's not feasible to provide all of these back to the user for the user to decide.

Reply:
Assuming we do continue to use time range as the validity check, I'd prefer `now + 1 year` as the top of the range, to `Jan 1, 2025`. It's more likely that an analyst will be using osquery in 2025 than that they have a system clock 1 year in the past.